util: fix compilation with mingw-w64 7.0.0

fanquake opened the pull request:
Something has changed in the mingw-w64 headers such that we
no-longer compile when using 7.0.0.

```bash
util/time.cpp: In function 'std::__cxx11::string FormatISO8601DateTime(int64_t)':
util/time.cpp:84:9: error: 'gmtime_r' was not declared in this scope
     if (gmtime_r(&time_val, &ts) == nullptr) {
         ^~~~~~~~
util/time.cpp: In function 'std::__cxx11::string FormatISO8601Date(int64_t)':
util/time.cpp:97:9: error: 'gmtime_r' was not declared in this scope
     if (gmtime_r(&time_val, &ts) == nullptr) {
```

Looking at time.h, it seems that `gmtime_r()` is only available when
`_POSIX_C_SOURCE` is defined. This must have been the case for 6.0.0 
(which we compile fine using), but no-longer seems to be for 7.0.0?

I've checked that adding `-D_POSIX_C_SOURCE=200112L` to our compile
flags does fix the issue above.

However, an alternative solution seems to be to just use `gmtime_s()`
instead, when compiling with `mingw-w64`, as `gmtime_r()` [just wraps
`gmtime_s()` anyways](https://github.com/mirror/mingw-w64/blob/7c03b11bf173944679102bbe0ac061842e2f594b/mingw-w64-headers/crt/time.h#L284).

I've tested this change crosss-compiling on Debian Bullseye ([mingw-w64 7.0.0](https://packages.debian.org/source/bullseye/mingw-w64)) 
and Buster ([mingw-w64 6.0.0](https://packages.debian.org/source/buster/mingw-w64)).

hebasto commented:
Concept ACK.

hebasto commented:
> However, an alternative solution seems to be to just use `gmtime_s()`
> instead, when compiling with `mingw-w64`, as `gmtime_r()` [just wraps `gmtime_s()` anyways](https://github.com/mirror/mingw-w64/blob/7c03b11bf173944679102bbe0ac061842e2f594b/mingw-w64-headers/crt/time.h#L284).

Does this alternative solution has any downsides?

practicalswift commented:
Concept ACK

dongcarl commented:
Did a bit more digging. Here's what I found:

[This commmit](https://github.com/mirror/mingw-w64/commit/69effce76945e75711e8e3fefaf3a8826ef7a66d) is the one that caused the change in behaviour.

There's a very short mailing list thread about it [here](https://sourceforge.net/p/mingw-w64/mailman/message/36638299/).

My conversation with the author of the commit on IRC:
```
15:02 <dongcarl> jacekc: After your "Don't define _POSIX_THREAD_SAFE_FUNCTIONS." patch, which landed in 7.0.0, what is the right way to get `gmtime_r` in my application?
15:02 * dongcarl waits patiently
15:04 <jacekc> dongcarl: ideally, I'd suggest to use gmtime_s
15:04 <dongcarl> jacekc: Why's that?
15:05 <jacekc> because that's what the platform you target offers
15:06 <jacekc> but if you prefer mingw wrappers, you may define _POSIX_THREAD_SAFE_FUNCTIONS or _POSIX_C_SOURCE
15:07 <dongcarl> jacekc: I think I see what you're saying. I'm assuming the mingw wrappers don't offer thread safety, even though the name is gmtime_r?
15:07 * dongcarl is thankful for the patient response
15:09 <jacekc> both _s and _r variants offer thread safety
15:09 <jacekc> _r variant is just an inline wrapper in time.h aroud _s
15:12 <dongcarl> jacekc: I see. Out of curiosity, what's the difference between defining _POSIX_THREAD_SAFE_FUNCTIONS and _POSIX_C_SOURCE? Feel free to just send me a link to read
15:14 <jacekc> _POSIX_THREAD_SAFE_FUNCTIONS enables only a few inline functions in time.h, _POSIX_C_SOURCE enables much more (most notably mongw stdio implementation)
15:15 <jacekc> I don't know if it's documented anywhere, but it's easy to find in the sources
15:16 <dongcarl> jacekc: Thanks for your help. Very much appreciated and have a tremendous day!
15:16 <jacekc> you welcome
```

Here's some more general use information on [_POSIX_C_SOURCE](https://perma.cc/NTK8-7R2Q) and [_POSIX_THREAD_SAFE_FUNCTIONS](https://perma.cc/ADU7-UQLK).

MarcoFalke commented:
Looks like this makes the include order of headers more resilient and fixes a bug I was running in to a year ago: https://github.com/bitcoin/bitcoin/pull/16046/files#r285601479

If you feel like, you might sort the includes in the header now:

```diff
diff --git a/src/util/time.h b/src/util/time.h
index 77de1e047d..a8c06ef350 100644
--- a/src/util/time.h
+++ b/src/util/time.h
@@ -6,9 +6,9 @@
 #ifndef BITCOIN_UTIL_TIME_H
 #define BITCOIN_UTIL_TIME_H
 
+#include <chrono>
 #include <stdint.h>
 #include <string>
-#include <chrono>
 
 void UninterruptibleSleep(const std::chrono::microseconds& n);

dongcarl commented:
If I'm thinking about this the right way, your diff implies that previously the `#ifdef _MSC_VER` would fail, and branch to the `#else`, which used `gmtime_r`, which would be fine because it was defined by default in mingw-w64.

If `#ifdef _MSC_VER` is how we detect if we're building for windows, and it's failing, then perhaps there are other places in the codebase where this detection needs to be fixed?

fanquake commented:
> If you feel like, you might sort the includes in the header now:

Done.

> If #ifdef _MSC_VER is how we detect if we're building for windows, and it's failing, then perhaps there are other places in the codebase where this detection needs to be fixed?

I've looked into this, and there are some cases where we are using `_MSC_VER` where we  might be running into unexpected behavior. I don't necessarily want to address that as part of this PR, but will definitely be following up.

maflcko reviewed (approved):
ACK 1b801e770477e756d67402970694bca73fa0ac7f

cc @sipsorcery

hebasto commented:
> Something has changed in the mingw-w64 headers such that we
> no-longer compile when using 7.0.0.

Interesting why master could be built successfully with Ubuntu's [g++-mingw-w64 7.3.0](https://packages.ubuntu.com/bionic/g++-mingw-w64)?

fanquake commented:
> Interesting why master could be built successfully with Ubuntu's g++-mingw-w64 7.3.0?

The version of `mingw-w64` being used is [5.0.0](https://packages.ubuntu.com/bionic/mingw-w64). `7.3.0` is the version of GCC.

hebasto reviewed (approved):
ACK 1b801e770477e756d67402970694bca73fa0ac7f, tested on Ubuntu 20.04 with [mingw-w64 7.0.0](https://packages.ubuntu.com/focal/mingw-w64).

dongcarl commented:
From #bitcoin-builds meeting notes today: We should investigate how to detect features (i.e. check for function availability at configure time), rather than detecting platforms. This might be more robust.

MarcoFalke commented:
@dongcarl Is this a general advise/issue or a blocker on this specific patch set?

Empact commented:
This seems possibly problematic due to:
> The implementation of gmtime_s in Microsoft CRT is incompatible with the C standard since it has reversed parameter order.
https://en.cppreference.com/w/c/chrono/gmtime

dongcarl commented:
@MarcoFalke I think it's a blocker while we investigate if feature detection is the right path take going forward, no point merging this then changing right after (unless the release schedule dictates?)

Empact committed:
build: Detect gmtime_* definitions via configure

This improves the portability of the codebase and fixes compilation
with mingw-w64 7.0+.

Co-authored-by: fanquake <fanquake@gmail.com>

Empact commented:
Here's an alternative: https://github.com/bitcoin/bitcoin/compare/master...Empact:2020-03-gmtime-config

Note I haven't run it on windows/mingw, and I'm not sure how to get the test to distinguish between MSC-style `gmtime_s` and C-standard-style `gmtime_s`.

sipsorcery commented:
ACK 1b801e770477e756d67402970694bca73fa0ac7f.

laanwj commented:
Using the windows function on windows makes sense, imo.

laanwj commented on src/util/time.cpp:
```diff
@@ -78,7 +78,7 @@ int64_t GetSystemTimeInSeconds()
     struct tm ts;
     time_t time_val = nTime;
-#ifdef _MSC_VER
+#if defined(_MSC_VER) || defined(__MINGW64_VERSION_MAJOR)
```
can't we change this to `#ifdef WIN32` ?

dongcarl commented:
Hmmm, I think perhaps what we want to do is similar to what's being done here: https://github.com/maru/libmicrohttpd-http2/blob/316af7c06537e524c6ef02bc1d5f59b39af7d1d1/configure.ac#L1033-L1082

It's feature-based detection, and we'll have clarity between `HAVE_C11_GMTIME_S`, `HAVE_W32_GMTIME_S`, and `HAVE_GMTIME_R`.

laanwj commented:
Yeah, that's another way.
I wish this kind of basic functionality was standardized by now. This comes up so often.

-----

If we still want this for 0.20, we need to make a decision here quickly though. I'm sorry for starting a bikeshed :laughing:

dongcarl commented:
Yeah I think we should go with @Empact's https://github.com/bitcoin/bitcoin/compare/master...Empact:2020-03-gmtime-config, it's feature-based detection, and errors at configure-time if we don't support the platform instead of error-ing at make-time.

Perhaps the error message should read "both gmtime_r and MSC-style gmtime_s are missing" for disambiguation.

Ping @fanquake

-----

I think we should get this into 0.20, because Ubuntu 20.04 LTS is going to release with mingw-w64 v7.0.0 (also, would help unblock https://github.com/bitcoin/bitcoin/pull/17595#discussion_r401312121)

fanquake commented:
> Yeah I think we should go with @Empact's

I'm taking a look at this, and it doesn't quite work as expected. The `AC_MSG_CHECKING` output doesn't actually get printed in the `gmtime_s` success case. However I think we can use this once that is fixed and messages are adjusted as required.

fanquake commented:
> doesn't actually get printed in the gmtime_s success case.

Maybe I was just imaging this.. Have pushed up some new changes.

Tested on macOS:
macOS
```bash
checking for gmtime_r... yes
```

mingw-w64 cross compile:
```bash
checking for gmtime_r... no
checking for gmtime_s... yes
```

debian:
```bash
checking for gmtime_r... yes
```

MSVC ✅

Empact commented:
Note one thing I have not tested with this implementation is how it operates if gmtime_r is not present, but C11-style gmtime_s is.

hebasto reviewed (commented):
Tested a46484c8b3715725f5dc0b8ad1bf921880ed9af1 on focal and bionic. Both produce:
```
$ CONFIG_SITE=$PWD/depends/x86_64-w64-mingw32/share/config.site ./configure
...
checking for gmtime_r... no
checking for gmtime_s... yes
...
```

To make things clear about cross compiling on systems with pre-7.0 mingw-w64 headers, e.g., on bionic:
on master the `gmtime_r()` is used, and with this PR the `gmtime_s()` is used, right?

laanwj commented:
ACK a46484c8b3715725f5dc0b8ad1bf921880ed9af1

theuni commented:
Post-merge ACK. Thanks for making this test-based, I think this is a nice improvement.

fanquake commented:
> To make things clear about cross compiling on systems with pre-7.0 mingw-w64 headers, e.g., on bionic:

@hebasto the breakdown is:

| commit | Debian Buster ([mingw-w64 6.0.0](https://packages.debian.org/source/buster/mingw-w64)) | Debian Bullseye ([mingw-w64 7.0.0](https://packages.debian.org/source/bullseye/mingw-w64)) |
| - | - | - |
| 5c1ba3a10a18c6edee20f27f3609ffbbbaa2e8f3~1 | gmtime_r() | fails to compile |
| 5c1ba3a10a18c6edee20f27f3609ffbbbaa2e8f3 | gmtime_s() | gmtime_s() |
